File: src/geometry_angle_relationships_medium.py
"""
geometry - angle_relationships (medium)
Generated: 2026-02-22T04:03:42.217565
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = choice(['vertical_angles', 'linear_pair', 'complementary', 'supplementary', 'angle_sum'])
    
    if problem_type == 'vertical_angles':
        # Vertical angles problem with algebraic expressions
        coeff1 = randint(2, 5)
        const1 = randint(5, 25)
        coeff2 = randint(2, 5)
        const2 = randint(5, 25)
        
        # Ensure different coefficients
        while coeff1 == coeff2:
            coeff2 = randint(2, 5)
        
        # Solve for x: coeff1*x + const1 = coeff2*x + const2
        x_val = (const2 - const1) / (coeff1 - coeff2)
        
        # Make sure x is a positive integer
        if x_val <= 0 or x_val != int(x_val):
            # Adjust to get clean solution
            const2 = const1 + (coeff1 - coeff2) * randint(2, 8)
            x_val = (const2 - const1) // (coeff1 - coeff2)
        
        angle_measure = coeff1 * x_val + const1
        
        expr1 = f"{coeff1}x + {const1}" if const1 > 0 else f"{coeff1}x - {-const1}"
        expr2 = f"{coeff2}x + {const2}" if const2 > 0 else f"{coeff2}x - {-const2}"
        
        d = Diagram(width=300, height=250)
        # Draw two intersecting lines
        d.line((-3, -2), (3, 2))
        d.line((-2, 2), (2, -2))
        # Label angles
        d.text(1.5, 0.8, f"${expr1}°$")
        d.text(-1.5, -0.8, f"${expr2}°$")
        
        question = f"Two lines intersect forming vertical angles. One angle measures $({expr1})°$ and its vertical angle measures $({expr2})°$. Find the value of $x$."
        
        return problem(
            question=question,
            answer=int(x_val),
            difficulty=(1300, 1450),
            topic="geometry/angle_relationships",
            solution=steps(
                f"Vertical angles are congruent, so we set the expressions equal:",
                f"${expr1} = {expr2}$",
                f"${coeff1}x - {coeff2}x = {const2} - {const1}$",
                f"${coeff1 - coeff2}x = {const2 - const1}$",
                f"$x = {int(x_val)}$"
            ),
            image=d.render()
        )
    
    elif problem_type == 'linear_pair':
        # Linear pair (supplementary adjacent angles)
        coeff = randint(2, 4)
        const = randint(10, 30)
        
        # Solve: (coeff*x + const) + (90 - x) = 180
        # coeff*x - x = 180 - 90 - const
        x_val = (90 - const) // (coeff - 1)
        
        # Ensure positive integer
        if x_val <= 0:
            const = randint(5, 20)
            x_val = (90 - const) // (coeff - 1)
        
        angle1 = coeff * x_val + const
        angle2 = 180 - angle1
        
        expr1 = f"{coeff}x + {const}" if const > 0 else f"{coeff}x - {-const}"
        
        d = Diagram(width=300, height=200)
        # Draw a line with two adjacent angles
        d.line((-3, 0), (3, 0))
        d.line((0, 0), (1.5, 2))
        d.angle_arc((0, 0), (3, 0), (1.5, 2), label=f"${expr1}°$")
        d.angle_arc((0, 0), (1.5, 2), (-3, 0), label=f"${angle2}°$")
        
        question = f"Two angles form a linear pair. One angle measures $({expr1})°$ and the other measures ${angle2}°$. Find the value of $x$."
        
        return problem(
            question=question,
            answer=int(x_val),
            difficulty=(1350, 1500),
            topic="geometry/angle_relationships",
            solution=steps(
                f"Linear pairs are supplementary (sum to $180°$):",
                f"$({expr1}) + {angle2} = 180$",
                f"${coeff}x + {const + angle2} = 180$",
                f"${coeff}x = {180 - const - angle2}$",
                f"$x = {int(x_val)}$"
            ),
            image=d.render()
        )
    
    elif problem_type == 'complementary':
        # Complementary angles
        coeff = randint(2, 5)
        const = randint(5, 15)
        
        # Solve: (coeff*x + const) + 2x = 90
        x_val = (90 - const) // (coeff + 2)
        
        # Ensure clean solution
        if x_val <= 0:
            const = randint(3, 10)
            x_val = (90 - const) // (coeff + 2)
        
        angle1 = coeff * x_val + const
        angle2 = 90 - angle1
        
        expr1 = f"{coeff}x + {const}"
        
        question = f"Two complementary angles are given. One angle measures $({expr1})°$ and the other measures $2x°$. Find the value of $x$."
        
        return problem(
            question=question,
            answer=int(x_val),
            difficulty=(1300, 1450),
            topic="geometry/angle_relationships",
            solution=steps(
                f"Complementary angles sum to $90°$:",
                f"$({expr1}) + 2x = 90$",
                f"${coeff + 2}x + {const} = 90$",
                f"${coeff + 2}x = {90 - const}$",
                f"$x = {int(x_val)}$"
            )
        )
    
    elif problem_type == 'supplementary':
        # Supplementary angles
        coeff1 = randint(3, 6)
        coeff2 = randint(2, 4)
        const = randint(8, 20)
        
        # Solve: (coeff1*x - const) + (coeff2*x + const) = 180
        x_val = 180 // (coeff1 + coeff2)
        
        angle1 = coeff1 * x_val - const
        angle2 = coeff2 * x_val + const
        
        expr1 = f"{coeff1}x - {const}"
        expr2 = f"{coeff2}x + {const}"
        
        question = f"Two supplementary angles measure $({expr1})°$ and $({expr2})°$. Find the value of $x$."
        
        return problem(
            question=question,
            answer=int(x_val),
            difficulty=(1400, 1550),
            topic="geometry/angle_relationships",
            solution=steps(
                f"Supplementary angles sum to $180°$:",
                f"$({expr1}) + ({expr2}) = 180$",
                f"${coeff1}x - {const} + {coeff2}x + {const} = 180$",
                f"${coeff1 + coeff2}x = 180$",
                f"$x = {int(x_val)}$"
            )
        )
    
    else:  # angle_sum
        # Angles around a point or in a triangle
        if choice([True, False]):
            # Triangle angle sum
            coeff = randint(2, 4)
            angle1 = randint(40, 60)
            angle2 = randint(30, 50)
            
            # Third angle: coeff*x
            x_val = (180 - angle1 - angle2) // coeff
            
            question = f"In a triangle, two angles measure ${angle1}°$ and ${angle2}°$. The third angle measures ${coeff}x°$. Find the value of $x$."
            
            return problem(
                question=question,
                answer=int(x_val),
                difficulty=(1300, 1450),
                topic="geometry/angle_relationships",
                solution=steps(
                    f"The sum of angles in a triangle is $180°$:",
                    f"${angle1} + {angle2} + {coeff}x = 180$",
                    f"${angle1 + angle2} + {coeff}x = 180$",
                    f"${coeff}x = {180 - angle1 - angle2}$",
                    f"$x = {int(x_val)}$"
                )
            )
        else:
            # Angles around a point
            coeff1 = randint(2, 4)
            coeff2 = randint(2, 4)
            const1 = randint(10, 30)
            const2 = randint(10, 30)
            
            # Three angles: (coeff1*x + const1), (coeff2*x + const2), and remaining
            # Sum = 360
            x_val = (360 - const1 - const2 - 90) // (coeff1 + coeff2)
            
            expr1 = f"{coeff1}x + {const1}"
            expr2 = f"{coeff2}x + {const2}"
            
            question = f"Three angles around a point measure $({expr1})°$, $({expr2})°$, and $90°$. Find the value of $x$."
            
            return problem(
                question=question,
                answer=int(x_val),
                difficulty=(1450, 1600),
                topic="geometry/angle_relationships",
                solution=steps(
                    f"Angles around a point sum to $360°$:",
                    f"$({expr1}) + ({expr2}) + 90 = 360$",
                    f"${coeff1 + coeff2}x + {const1 + const2 + 90} = 360$",
                    f"${coeff1 + coeff2}x = {360 - const1 - const2 - 90}$",
                    f"$x = {int(x_val)}$"
                )
            )

emit(generate())

File: src/geometry_angle_relationships_medium.jl
# geometry - angle_relationships (medium)
# Generated: 2026-03-08T20:01:54.954045

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/angle_relationships")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Complementary angles problem (1200-1400)
        x_val = randint(15, 75)
        angle1 = x_val
        angle2 = 90 - x_val
        
        # Create diagram
        d = DiagramObj(width=400, height=300)
        # Draw two complementary angles
        line!(d, 50, 250, 200, 250)  # horizontal base
        line!(d, 50, 250, 50, 100)   # vertical line
        line!(d, 50, 250, 150, 150)  # angle bisector
        
        angle_arc!(d, 50, 250, 40, 0, angle1, label="$(angle1)°")
        angle_arc!(d, 50, 250, 50, angle1, 90, label="$(angle2)°")
        
        text!(d, 200, 50, "Complementary Angles", fontsize=16)
        
        problem(
            question="Two angles are complementary. If one angle measures $(angle1)°, what is the measure of the other angle?",
            answer=angle2,
            difficulty=(1200, 1400),
            solution=steps(
                "Complementary angles sum to 90°",
                sol("Setup", "$(angle1)° + x = 90°"),
                sol("Solve", "x = 90° - $(angle1)° = $(angle2)°")
            ),
            image=render(d),
            time=60
        )
        
    elseif problem_type == 2
        # Supplementary angles with algebra (1300-1500)
        k = randint(2, 8)
        x_val = randint(10, 25)
        angle1_expr = k * x_val + randint(-20, 20)
        angle2 = 180 - angle1_expr
        
        # Create diagram
        d = DiagramObj(width=450, height=250)
        line!(d, 50, 150, 400, 150)
        line!(d, 225, 150, 300, 50)
        
        angle_arc!(d, 225, 150, 50, 0, 180 - atan((150-50)/(300-225))*180/π, label="$(angle1_expr)°")
        angle_arc!(d, 225, 150, 60, 180 - atan((150-50)/(300-225))*180/π, 180, label="$(angle2)°")
        
        text!(d, 200, 220, "Supplementary Angles", fontsize=16)
        
        problem(
            question="Two angles are supplementary. If one angle measures $(angle1_expr)°, find the measure of the other angle.",
            answer=angle2,
            difficulty=(1300, 1500),
            solution=steps(
                "Supplementary angles sum to 180°",
                sol("Setup", "$(angle1_expr)° + x = 180°"),
                sol("Solve", "x = 180° - $(angle1_expr)° = $(angle2)°")
            ),
            image=render(d),
            time=75
        )
        
    elseif problem_type == 3
        # Vertical angles (1200-1400)
        angle_measure = randint(35, 145)
        
        # Create diagram with intersecting lines
        d = DiagramObj(width=400, height=400)
        line!(d, 50, 200, 350, 200)  # horizontal
        line!(d, 120, 50, 280, 350)  # diagonal
        
        point!(d, 200, 200, label="", size=4)
        
        angle_arc!(d, 200, 200, 40, 0, atan((200-50)/(280-120))*180/π, label="$(angle_measure)°")
        angle_arc!(d, 200, 200, 40, 180, 180 + atan((200-50)/(280-120))*180/π, label="x")
        
        text!(d, 200, 380, "Find x (vertical angles)", fontsize=16)
        
        problem(
            question="Two lines intersect forming vertical angles. If one angle measures $(angle_measure)°, find the measure of its vertical angle x.",
            answer=angle_measure,
            difficulty=(1200, 1400),
            solution=steps(
                "Vertical angles are congruent",
                sol("Property", "Vertical angles have equal measures"),
                sol("Answer", "x = $(angle_measure)°")
            ),
            image=render(d),
            time=60
        )
        
    elseif problem_type == 4
        # Linear pair with algebra (1400-1600)
        coef = randint(2, 6)
        x_val = randint(15, 30)
        const_term = randint(-15, 15)
        
        angle1 = coef * x_val + const_term
        angle2 = 180 - angle1
        
        # Create diagram
        d = DiagramObj(width=450, height=250)
        line!(d, 50, 150, 400, 150)
        line!(d, 225, 150, 300, 50)
        
        angle_arc!(d, 225, 150, 50, 0, atan((150-50)/(300-225))*180/π, label="($(coef)x$(const_term >= 0 ? "+" : "")$(const_term))°")
        angle_arc!(d, 225, 150, 60, atan((150-50)/(300-225))*180/π, 180, label="$(angle2)°")
        
        text!(d, 180, 220, "Linear Pair - Find x", fontsize=16)
        
        problem(
            question="Two angles form a linear pair. One angle measures ($(coef)x $(const_term >= 0 ? "+" : "") $(const_term))° and the other measures $(angle2)°. Find the value of x.",
            answer=x_val,
            difficulty=(1400, 1600),
            solution=steps(
                "Linear pairs sum to 180°",
                sol("Equation", "($(coef)x $(const_term >= 0 ? "+" : "") $(const_term)) + $(angle2) = 180"),
                sol("Simplify", "$(coef)x = $(180 - angle2 - const_term)"),
                sol("Answer", "x = $(x_val)")
            ),
            image=render(d),
            time=90
        )
        
    else
        # Angles around a point (1500-1700)
        x_val = randint(12, 35)
        angle1 = randint(40, 80)
        angle2 = randint(50, 90)
        coef1 = randint(2, 4)
        coef2 = randint(2, 4)
        
        angle3 = coef1 * x_val
        angle4 = coef2 * x_val
        
        # Adjust to make sure sum is 360
        total_known = angle1 + angle2
        total_var = coef1 + coef2
        x_val = (360 - total_known) ÷ total_var
        
        angle3 = coef1 * x_val
        angle4 = coef2 * x_val
        
        # Create diagram
        d = DiagramObj(width=400, height=400)
        point!(d, 200, 200, label="", size=5)
        
        # Draw four rays from center
        line!(d, 200, 200, 350, 200)  # 0°
        line!(d, 200, 200, 280, 80)   # angle1
        line!(d, 200, 200, 120, 120)  # angle1 + angle2
        line!(d, 200, 200, 120, 280)  # remaining
        
        angle_arc!(d, 200, 200, 40, 0, 40, label="$(angle1)°")
        angle_arc!(d, 200, 200, 50, 50, 110, label="$(angle2)°")
        angle_arc!(d, 200, 200, 45, 120, 180, label="$(coef1)x°")
        angle_arc!(d, 200, 200, 55, 190, 300, label="$(coef2)x°")
        
        text!(d, 200, 380, "Angles around a point - Find x", fontsize=16)
        
        problem(
            question="Four angles meet at a point. The angles measure $(angle1)°, $(angle2)°, $(coef1)x°, and $(coef2)x°. Find the value of x.",
            answer=x_val,
            difficulty=(1500, 1700),
            solution=steps(
                "Angles around a point sum to 360°",
                sol("Equation", "$(angle1) + $(angle2) + $(coef1)x + $(coef2)x = 360"),
                sol("Simplify", "$(angle1 + angle2) + $(coef1 + coef2)x = 360"),
                sol("Solve", "$(coef1 + coef2)x = $(360 - angle1 - angle2)"),
                sol("Answer", "x = $(x_val)")
            ),
            image=render(d),
            time=120
        )
    end
end